community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_1/TransactionBoundQueryContext.scala

```diff
@@ -583,26 +584,60 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
     pathFinder.findAllPaths(left, right).iterator().asScala
   }
 
-  override def callReadOnlyProcedure(name: QualifiedProcedureName, args: Seq[Any]) =
-    callProcedure(name, args, transactionalContext.statement.readOperations().procedureCallRead)
+  override def callReadOnlyProcedure(name: QualifiedProcedureName, args: Seq[Any], allowed: String) = {
+    val revertable = transactionalContext.accessMode match {
+      case a: AuthSubject if a.hasRole(allowed) =>
+        Some(transactionalContext.restrictCurrentTransaction(AccessMode.Static.OVERRIDE_READ))
+      case _ => None
+    }
+    callProcedure(name, args, transactionalContext.statement.readOperations().procedureCallRead, revertable.foreach(_.close))
+  }
 
-  override def callReadWriteProcedure(name: QualifiedProcedureName, args: Seq[Any]) =
-    callProcedure(name, args, transactionalContext.statement.dataWriteOperations().procedureCallWrite)
+  override def callReadWriteProcedure(name: QualifiedProcedureName, args: Seq[Any], allowed: String) = {
+    val revertable = transactionalContext.accessMode match {
+      case a: AuthSubject if a.hasRole(allowed) =>
+        Some(transactionalContext.restrictCurrentTransaction(AccessMode.Static.OVERRIDE_WRITE))
+      case _ => None
+    }
+    callProcedure(name, args, transactionalContext.statement.dataWriteOperations().procedureCallWrite,
+                  revertable.foreach(_.close))
+  }
 
-  override def callSchemaWriteProcedure(name: QualifiedProcedureName, args: Seq[Any]) =
-    callProcedure(name, args, transactionalContext.statement.schemaWriteOperations().procedureCallSchema)
 
-  override def callDbmsProcedure(name: QualifiedProcedureName, args: Seq[Any]) =
-    callProcedure(name, args, transactionalContext.dbmsOperations.procedureCallDbms(_, _))
+  override def callSchemaWriteProcedure(name: QualifiedProcedureName, args: Seq[Any], allowed: String) = {
+    val revertable = transactionalContext.accessMode match {
+      case a: AuthSubject if a.hasRole(allowed) =>
+        Some(transactionalContext.restrictCurrentTransaction(AccessMode.Static.OVERRIDE_SCHEMA))
+      case _ => None
+    }
+    callProcedure(name, args, transactionalContext.statement.schemaWriteOperations().procedureCallSchema, revertable.foreach(_.close))
+  }
+
+  override def callDbmsProcedure(name: QualifiedProcedureName, args: Seq[Any], allowed: String) = {
+    val revertable = transactionalContext.accessMode match {
+      case a: AuthSubject if a.hasRole(allowed) =>
+        Some(transactionalContext.restrictCurrentTransaction(AccessMode.Static.OVERRIDE_SCHEMA))
+      case _ => None
+    }
+    callProcedure(name, args, transactionalContext.dbmsOperations.procedureCallDbms(_, _), revertable.foreach(_.close))
```

Comment:
I realize we need to specify what `allowed` should to for DBMS procedures. Currently the procedures enforce access rights themselves per procedure, so we should probably pass the `allowed` string into procedureCallDbms here so that it can be injected into the procedure context.